src/disco/shred/fd_stake_ci.c

```diff
@@ -0,0 +1,258 @@
+#include "fd_stake_ci.h"
+#include "../../util/net/fd_ip4.h" /* Just for debug */
+
+#define SORT_NAME sort_pubkey
+#define SORT_KEY_T fd_shred_dest_weighted_t
+#define SORT_BEFORE(a,b) memcmp( (a).pubkey.uc, (b).pubkey.uc, 32UL )
+#include "../../util/tmpl/fd_sort.c"
+
+void *
+fd_stake_ci_new( void             * mem,
+                fd_pubkey_t const * identity_key ) {
+  fd_stake_ci_t * info = (fd_stake_ci_t *)mem;
+
+  if( FD_UNLIKELY( MAX_SHRED_DEST_FOOTPRINT != fd_shred_dest_footprint( MAX_SHRED_DESTS ) ) )
+    FD_LOG_ERR(( "MAX_SHRED_DEST_FOOTPRINT should be set to %lu", fd_shred_dest_footprint( MAX_SHRED_DESTS ) ));
+
+
+  fd_stake_weight_t dummy_stakes[ 1 ] = {{ .key = {{0}}, .stake = 1UL }};
+  fd_shred_dest_weighted_t dummy_dests[ 1 ] = {{ .pubkey = *identity_key }};
+
+  /* Initialize first 2 to satisfy invariants */
+  info->stake_weight[ 0 ] = dummy_stakes[ 0 ];
+  info->shred_dest  [ 0 ] = dummy_dests [ 0 ];
+  for( ulong i=0UL; i<2UL; i++ ) {
+    fd_per_epoch_info_t * ei = info->epoch_info + i;
+    ei->epoch      = i;
+    ei->start_slot = 0UL;
+    ei->slot_cnt   = 0UL;
+
+    ei->lsched = fd_epoch_leaders_join( fd_epoch_leaders_new( ei->_lsched, 0UL, 0UL, 1UL, 1UL,    info->stake_weight       ) );
+    ei->sdest  = fd_shred_dest_join   ( fd_shred_dest_new   ( ei->_sdest,  info->shred_dest, 1UL, ei->lsched, identity_key ) );
+  }
+  info->identity_key[ 0 ] = *identity_key;
+
+  return (void *)info;
+}
+
+fd_stake_ci_t * fd_stake_ci_join( void * mem ) { return (fd_stake_ci_t *)mem; }
+
+void * fd_stake_ci_leave ( fd_stake_ci_t * info ) { return (void *)info; }
+void * fd_stake_ci_delete( void          * mem  ) { return mem;          }
+
+
+void
+fd_stake_ci_stake_msg_init( fd_stake_ci_t * info,
+                            uchar const   * new_message ) {
+  ulong const * hdr = fd_type_pun_const( new_message );
+
+  ulong epoch               = hdr[ 0 ];
+  ulong staked_cnt          = hdr[ 1 ];
+  ulong start_slot          = hdr[ 2 ];
+  ulong slot_cnt            = hdr[ 3 ];
+
+  /* TODO: double check dest_cnt */
+
+  info->scratch->epoch      = epoch;
+  info->scratch->start_slot = start_slot;
+  info->scratch->slot_cnt   = slot_cnt;
+  info->scratch->staked_cnt = staked_cnt;
+
+  fd_memcpy( info->stake_weight, hdr+4UL, sizeof(fd_stake_weight_t)*staked_cnt );
+}
+
+static inline void
+log_summary( char const * msg, fd_stake_ci_t * info ) {
```

Comment:
Do we want this on NOTICE in prod?

Reply:
I removed the one on lookup failure, which could be triggered by someone sending a malicious shred. I'll leave the other ones for now, since they happen only around epoch transitions or contact info changes (capped at 1x per 5 seconds). Solana Labs does something similar, and it has been useful for debugging.